primitives/runtime/src/offchain/storage.rs

```diff
@@ -28,6 +28,26 @@ pub struct StorageValueRef<'a> {
 	kind: StorageKind,
 }
 
+/// Reason for not being able to provide the stored value
+#[derive(Debug, PartialEq, Eq)]
+pub enum StorageRetrivalError {
+	/// Value found but undecodable
+	Undecodable,
+}
+
+/// Possible errors when mutating a storage value.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MutateStorageError<T, E> {
+	/// The underlying db failed to update due to a concurrent modification.
+	/// Contains the new value that was not stored.
+	CompareAndSetFailed(T),
+	/// The function given to us to create the value to be stored failed.
+	/// May be used to signal that having looked at the existing value,
+	/// they don't want to mutate it.
+	ValueFunctionFailed(E)
```

Comment:
Nit: we could save a few characters by naming these `CompareAndSet`, `ValueFunction`. `Failed` feels redundant on an `Error` type.

Reply:
Am tempted to rename CompareAndSetFailed as ConcurrentModification because that's the reason why the compare and set fails.

Reply:
I'm almost tempted with renaming ValueFunctionFailed to MutationCancelled. 
(That is after all what we seem to use it for.)